File: JavaScript/selection_sort.js
// Function which sorts the given array (Selection sort)
function selectionSort(arr) {
    let lastIndex, maxIndex, temp;
    for (let i = 0; i < arr.length; i++) {
        lastIndex = arr.length - i - 1;  // Stores the last index element for a particular iteration
        maxIndex = 0;  // Stores the index of the largest element
        for (let j = 1; j < arr.length - i; j++) {  
            if (arr[j] > arr[maxIndex]) {  // If current element is greater than element at maxIndex position,
                maxIndex = j;              // then maxIndex = current index (j)
            }
        }
        // Swap largest element (of the particular iteration) 
        // with element at last index (of the particular iteration)
        temp = arr[maxIndex];  
        arr[maxIndex] = arr[lastIndex];
        arr[lastIndex] = temp;
    }
}

function main() {
    // Creating an unsorted array
    let arr = [10, -12, 57, 13, 41, -98];
    console.log("Entered array (unsorted) is:-");
    console.log("[" + arr.join(", ") + "]\n");
    selectionSort(arr);
    console.log("Displaying sorted array:-");
    // Displaying sorted array with proper formatting
    console.log("[" + arr.join(", ") + "]");
}

main();

/*
---Details---
Selection Sort Algorithm in JavaScript
Input: Unsorted array of integers
Output: Sorted array in ascending order

---Sample Inputs and Outputs---

---Sample Input - 1
Input array: [10, -12, 57, 13, 41, -98]
Output:- 
Entered array (unsorted) is:-
[10, -12, 57, 13, 41, -98]

Displaying sorted array:-
[-98, -12, 10, 13, 41, 57]

---Sample Input - 2
Input array: [19, -19, 0, 13, 51, 2, 76, 134, 12]
Output:- 
Entered array (unsorted) is:-
[19, -19, 0, 13, 51, 2, 76, 134, 12]

Displaying sorted array:-
[-19, 0, 2, 12, 13, 19, 51, 76, 134]

---Time Complexity---
Average Case, Best Case and Worst Case all are O(n^2)

---Space Complexity---
Space Complexity is O(1)
*/

File: Python/selectionSort.py
# Selection Sort
"""
Example 1:
Input:5 4 3 2 1
Output:[1, 2, 3, 4, 5]

Example 2:
Input:9 5 7 6 1 4 3 1
Output:[1, 1, 3, 4, 5, 6, 7, 9]

Example 3:
Input:2 3 7 5 4 6 8 1 9
Output:[1, 2, 3, 4, 5, 6, 7, 8, 9]
"""


def selectionSort(values):
    for i in range(len(values) - 1):
        for j in range(i + 1, len(values)):
            if values[i] > values[j]:
                values[i], values[j] = values[j], values[i]
    return values


print(selectionSort(list(map(int, input().split(" ")))))

# Time complexity: O(N^2)
# Space complexity: O(1)

